src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
-#define XMS_RESIZE_ANY_EXTENDED_MEMORY_BLOCK 0x8f
-
-#define	XMS_FUNCTION_NOT_IMPLEMENTED		0x80
-#define	HIGH_MEMORY_NOT_EXIST				0x90
-#define	HIGH_MEMORY_IN_USE					0x91
-#define	HIGH_MEMORY_NOT_ALLOCATED			0x93
-#define XMS_OUT_OF_SPACE					0xa0
-#define XMS_OUT_OF_HANDLES					0xa1
-#define XMS_INVALID_HANDLE					0xa2
-#define XMS_INVALID_SOURCE_HANDLE			0xa3
-#define XMS_INVALID_SOURCE_OFFSET			0xa4
-#define XMS_INVALID_DEST_HANDLE				0xa5
-#define XMS_INVALID_DEST_OFFSET				0xa6
-#define XMS_INVALID_LENGTH					0xa7
-#define XMS_BLOCK_NOT_LOCKED				0xaa
-#define XMS_BLOCK_LOCKED					0xab
-#define	UMB_ONLY_SMALLER_BLOCK				0xb0
-#define	UMB_NO_BLOCKS_AVAILABLE				0xb1
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+CHECK_NARROWING();
+
+// ***************************************************************************
+// Constants and type definitions
+// ***************************************************************************
+
+constexpr uint16_t XMS_VERSION        = 0x0300; // version 3.00
+constexpr uint16_t XMS_DRIVER_VERSION = 0x0301; // driver version 3.01
+
+// MS-DOS 6.22 defaults to 32 XMS handles, we can provide more without any
+// significant cost.
+constexpr uint8_t NumXmsHandles = 128;
+
+enum class Result : uint8_t {
+	OK                      = 0x00,
+	NOT_IMPLEMENTED         = 0x80,
+	VDISK_DETECTED          = 0x81, // not needed in DOSBox
+	A20_LINE_ERROR          = 0x82,
+	GENERAL_DRIVER_ERROR    = 0x8e, // not needed in DOSBox
+	HMA_NOT_EXIST           = 0x90,
+	HMA_IN_USE              = 0x91,
+	HMA_NOT_BIG_ENOUGH      = 0x92,
+	HMA_NOT_ALLOCATED       = 0x93,
+	A20_STILL_ENABLED       = 0x94,
+	XMS_OUT_OF_SPACE        = 0xa0,
+	XMS_OUT_OF_HANDLES      = 0xa1,
+	XMS_INVALID_HANDLE      = 0xa2,
+	XMS_INVALID_SRC_HANDLE  = 0xa3,
+	XMS_INVALID_SRC_OFFSET  = 0xa4,
+	XMS_INVALID_DST_HANDLE  = 0xa5,
+	XMS_INVALID_DST_OFFSET  = 0xa6,
+	XMS_INVALID_LENGTH      = 0xa7,
+	XMS_INVALID_OVERLAP     = 0xa8, // TODO: add support for this error
+	XMS_PARITY_ERROR        = 0xa9,
+	XMS_BLOCK_NOT_LOCKED    = 0xaa,
+	XMS_BLOCK_LOCKED        = 0xab,
+	XMS_LOCK_COUNT_OVERFLOW = 0xac,
+	XMS_LOCK_FAILED         = 0xad, // TODO: when should this be reported?
+	UMB_ONLY_SMALLER_BLOCK  = 0xb0,
+	UMB_NO_BLOCKS_AVAILABLE = 0xb1,
+	UMB_INVALID_SEGMENT     = 0xb2, // TODO: when should this be reported?
+};
 
 struct XMS_Block {
-	Bitu	size;
-	MemHandle mem;
-	uint8_t	locked;
-	bool	free;
+	uint32_t size_kb     = 0;
+	MemHandle mem_handle = -1;
+	// locked blocks should not be quietly moved by the XMS driver
+	uint8_t lock_count = 0;
+	bool is_free       = true;
 };
 
 #ifdef _MSC_VER
 #pragma pack (1)
 #endif
-struct XMS_MemMove{
-	uint32_t length;
-	uint16_t src_handle;
+struct XMS_MemMove {
+	uint32_t length = 0;
+
+	uint16_t src_handle = 0;
 	union {
 		RealPt realpt;
 		uint32_t offset;
 	} src;
-	uint16_t dest_handle;
+	uint16_t dst_handle = 0;
 	union {
 		RealPt realpt;
 		uint32_t offset;
-	} dest;
+	} dst;
```

Comment:
Fussy: I prefer src and dest, I thought was standard, but it's somewhat subjective 😄

Reply:
I'm seeing acronyms :-)

![2023-04-07_16-05](https://user-images.githubusercontent.com/1557255/230690439-790470cd-2455-47ea-81d4-d97f9117840d.png)

PST / MST / CST / ... -> all the daylight savings time zones in Canada.

Reply:
Yeah, I prefer to use these very sparingly: curr, prev, src, dest, maybe ptr, maybe addr, and that's about it. I'd use full names for anything else. Maybe some localised shortcuts are fine, e.g. within a function that deals with lots of register a _reg postfix might be okay. The more "public" a name is, the more "expanded" it needs to be, IMO.

Reply:
Or **D**enial-**O**f-**S**ervice-**Box** :D I like _dst_ for destination, because it has the same length as _src_ and the code often nicely aligns by itself.

But OK, changed to 'dest'.